automod/engine.go

```diff
@@ -0,0 +1,171 @@
+package automod
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"strings"
+
+	appbsky "github.com/bluesky-social/indigo/api/bsky"
+	"github.com/bluesky-social/indigo/atproto/identity"
+	"github.com/bluesky-social/indigo/atproto/syntax"
+	"github.com/bluesky-social/indigo/xrpc"
+)
+
+// runtime for executing rules, managing state, and recording moderation actions.
+//
+// TODO: careful when initializing: several fields should not be null or zero, even though they are pointer type.
+type Engine struct {
+	Logger    *slog.Logger
+	Directory identity.Directory
+	Rules     RuleSet
+	Counters  CountStore
+	Sets      SetStore
+	// used to persist moderation actions in mod service (optional)
+	AdminClient *xrpc.Client
+}
+
+func (e *Engine) ProcessIdentityEvent(ctx context.Context, t string, did syntax.DID) error {
+	// similar to an HTTP server, we want to recover any panics from rule execution
+	defer func() {
+		if r := recover(); r != nil {
+			e.Logger.Error("automod event execution exception", "err", r)
+		}
+	}()
+
+	ident, err := e.Directory.LookupDID(ctx, did)
+	if err != nil {
+		return fmt.Errorf("resolving identity: %w", err)
+	}
+	if ident == nil {
+		return fmt.Errorf("identity not found for did: %s", did.String())
+	}
+
+	evt := IdentityEvent{
+		Event{
+			Engine:  e,
+			Account: AccountMeta{Identity: ident},
+		},
+	}
+	if err := e.Rules.CallIdentityRules(&evt); err != nil {
+		return err
+	}
+	if evt.Err != nil {
+		return evt.Err
+	}
+	evt.CanonicalLogLine()
+	if err := evt.PersistAccountActions(ctx); err != nil {
+		return err
+	}
+	if err := evt.PersistCounters(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e *Engine) ProcessRecord(ctx context.Context, did syntax.DID, path, recCID string, rec any) error {
+	// similar to an HTTP server, we want to recover any panics from rule execution
+	defer func() {
+		if r := recover(); r != nil {
+			e.Logger.Error("automod event execution exception", "err", r)
+		}
+	}()
+
+	ident, err := e.Directory.LookupDID(ctx, did)
+	if err != nil {
+		return fmt.Errorf("resolving identity: %w", err)
+	}
+	if ident == nil {
+		return fmt.Errorf("identity not found for did: %s", did.String())
+	}
+	collection := strings.SplitN(path, "/", 2)[0]
+
+	switch collection {
```

Comment:
fyi: You can do a type switch in Go:
```go
switch rec := rec.(type) {
case *appbsky.FeedPost:
  // inside this case, rec is the correct type
}
```

Reply:
yeah, I had it that way first, but need to verify collection/type mapping somehow (aka, type needs to be FeedPost *and* collection is "app.bsky.feed.post") and this felt more legible than switch on type and verify collection string.